### ISSUE TITLE
error in wrapped code for c++11 enum class

### ISSUE THREAD

**Issue by kurzmeierm** [swig/swig#630]
Hi guys,
I want to build a perl module and using c++11 enum class. When I do the following, the values are not initialized. What am I doing wrong?

test.i

```
#ifdef SWIGPERL
    %module "test"
    %include <perl5.swg>
    %include <perl5/stl.i>
    %include <stl.i>
    %include <typemaps.i>
#endif
.
.
.
%inline %{
    enum Frequency1 {
      Val1_1=100,
      Val1_2,
      Val1_3,
      Val1_4
    };
    enum class Frequency2 {
      Val2_1=100,
      Val2_2,
      Val2_3,
      Val2_4
    };
%}
```

test.pl

``` perl
use strict;
use test;

print "Frequency1:(" . test::Val1_1 . ")\n";
print "Frequency2:(" . Frequency2::Val2_1() . ")\n";

exit 0;
```

output:
Frequency1:(100)
Frequency2:()

Frequency1 works as expected, Frequency2 would be generated but not correct initialized.
I've found the problem in the test_wrap.cpp file, the name attribute of function "get_sv" is not complete there is no class name of the enum (SWIG_prefix "Val2_1") should be (SWIG_prefix "Frequency2::Val2_1").

```
/*@SWIG:/home/michael/usr/share/swig/3.0.2/perl5/perltypemaps.swg,65,%set_constant@*/ do {
    SV *sv = get_sv((char*) SWIG_prefix "Val2_1", TRUE | 0x2 | GV_ADDMULTI);
    sv_setsv(sv, SWIG_From_int SWIG_PERL_CALL_ARGS_1(static_cast< int >(Frequency2::Val2_1)));
    SvREADONLY_on(sv);
  } while(0) /*@SWIG@*/;
```

But what I really want is:

test.hpp

```
    enum class Frequency2 {
      Val2_1=100,
      Val2_2,
      Val2_3,
      Val2_4
    };
```

test.i

```
#ifdef SWIGPERL
    %module "test"
    %include <perl5.swg>
    %include <perl5/stl.i>
    %include <stl.i>
    %include <typemaps.i>
#endif
.
.
.
%{
  #include <test.hpp>
%}

%{
    using Frequency2;
%}
enum class Frequency2 {
    Val2_1=100,
    Val2_2,
    Val2_3,
    Val2_4
  };
```

Does it only not run for perl or is it a general problem? Or what I hope: Do I something wrong?

Thanks for your support and best Regards,
Michael

**Comment by wsfulton** [comment]
I'm afraid that perl was not one of the target languages supported when C++11 enums were initially added, so a patch to add this in is needed.

**Comment by amanvm** [comment]
Hi @wsfulton 

Curious if there is any plan to support this in swig 4.0, swig 4.0 seems to be be leaning towards python. Also, what languages support this feature? I am porting a API to as many languages as I can using SWIG and would be good to know LCD of the features that work. For now I have replaced all enum classes with regular enums.

Btw, thanks for such a high-impact tool like SWIG.

Best,

Aman

**Comment by wsfulton** [comment]
Enum classes are part of c++11 and c++11 is currently tested using C#, Go, Java, Octave, Python, Ruby, R, Tcl. If I recall correctly, there was some issue with the Perl headers not being c++11 compatible so Perl got left behind. That may not be the case now though. I'd expect some c++11 features to thus not work in Perl, but most will as the majority of c++11 support comes from the common SWIG core.

I encourage anyone interested in using SWIG to contribute patches to missing features where they are required. This is the only way that SWIG will have missing features added as SWIG is an all volunteer community effort.

Incidentally, enum class support is currently sub-optimal for **all** the scripting languages as they are simply wrapped as a simple constant with the class name mangled into the constant name. A more optimal approach would be to wrap them into something akin to an enum class in the appropriate language, maybe a constant inside a class where there is no equivalent to an enum class.

**Comment by ojwb** [comment]
@wsfulton The Perl headers did indeed have some issues when compiling as C++11 (e.g. no space between a literal string and a macro) but they are C++11 clean now.  Perl 5.26 is dfefinitely OK.  Not sure which version it was actually fixed in, and the perldelta pages don't seem to explicitly say.

> swig 4.0 seems to be be leaning towards python-only updates

That really doesn't seem to be the case from a quick look at `CHANGES.current` on git master.  But if there is a bias towards a particular language it'll be because we are getting more contributions for that language than for others.

**Comment by ojwb** [comment]
I just tried to check if this was still present in git master.  The SWIG command line wasn't provided, but I think it must be:

```
swig -perl -c++ -const test.i
```

WIth that, the generated `test.pm` contains:

```
sub Val1_1 () { $testc::Val1_1 }
sub Val1_2 () { $testc::Val1_2 }
sub Val1_3 () { $testc::Val1_3 }
sub Val1_4 () { $testc::Val1_4 }
sub Frequency2::Val2_1 () { $testc::Frequency2::Val2_1 }
sub Frequency2::Val2_2 () { $testc::Frequency2::Val2_2 }
sub Frequency2::Val2_3 () { $testc::Frequency2::Val2_3 }
sub Frequency2::Val2_4 () { $testc::Frequency2::Val2_4 }
```

But the C level wrapping is `Frequency2_Val2_1`, etc:

```
    SV *sv = get_sv((char*) SWIG_prefix "Frequency2_Val2_1", TRUE | 0x2 | GV_ADDMULTI);
```

However, the `sub Frequency2::Val2_1 ()` part look like it doesn't work either.  I think we need to generate something like:

```
package test;

sub Val1_1 () { $testc::Val1_1 }
sub Val1_2 () { $testc::Val1_2 }
sub Val1_3 () { $testc::Val1_3 }
sub Val1_4 () { $testc::Val1_4 }

package test::Frequency2;

sub Val2_1 () { $testc::Frequency2_Val2_1 }
sub Val2_2 () { $testc::Frequency2_Val2_2 }
sub Val2_3 () { $testc::Frequency2_Val2_3 }
sub Val2_4 () { $testc::Frequency2_Val2_4 }
```

Manually modifying to that, the test script works if I add in a missing `test::` like so:

```
print "Frequency1:(" . test::Val1_1 . ")\n";
print "Frequency2:(" . test::Frequency2::Val2_1 . ")\n";
```

@talby Does the generated code just above look good to you?

**Comment by ojwb** [comment]
I didn't seem to explicitly say above, but this should all work with existing SWIG if you don't use `-const` - there's a `cpp11_strongly_typed_enumerations_runme.pl` which covers this case.

The first issue I noted above looks easy to fix - we're just using the wrong name here but it happens to work for traditional enums (and the non-`-const` cases around it already use _cmodule_`::`i_name_:

```
diff --git a/Source/Modules/perl5.cxx b/Source/Modules/perl5.cxx
index 85133a79..20fae81d 100644
--- a/Source/Modules/perl5.cxx
+++ b/Source/Modules/perl5.cxx
@@ -1117,7 +1117,7 @@ public:
               "tie %__", iname, "_hash,\"", is_shadow(type), "\", $",
               cmodule, "::", iname, ";\n", "$", iname, "= \\%__", iname, "_hash;\n", "bless $", iname, ", ", is_shadow(type), ";\n", NIL);
       } else if (do_constants) {
-       Printv(const_stubs, "sub ", name, " () { $", cmodule, "::", name, " }\n", NIL);
+       Printv(const_stubs, "sub ", name, " () { $", cmodule, "::", iname, " }\n", NIL);
        num_consts++;
       } else {
        Printv(var_stubs, "*", iname, " = *", cmodule, "::", iname, ";\n", NIL);
      Printv(const_stubs, "sub ", name, " () { $", cmodule, "::", iname, " }\n", NIL);
```

I'm running the testsuite and will push this fix assuming it passes.

Fixing the `sub` names looks fiddlier but I'll see if I can do that too.

**Comment by ojwb** [comment]
There are also problems with using the wrong name for a `%rename`-d enum (both in the enum class and "traditional" enum cases).  I have a patch which seems to fix both the above and this newly discovered wrinkle, but I still need to slot the reproducer into the testcase.

Generated Perl code is now:

```
package test;
sub PerlVal1_1 () { $testc::PerlVal1_1 }
sub Val1_2 () { $testc::Val1_2 }
sub Val1_3 () { $testc::Val1_3 }
sub Val1_4 () { $testc::Val1_4 }
package test::PerlFreq2;
sub PerlVal2_1 () { $testc::PerlFreq2_PerlVal2_1 }
sub Val2_2 () { $testc::PerlFreq2_Val2_2 }
sub Val2_3 () { $testc::PerlFreq2_Val2_3 }
sub Val2_4 () { $testc::PerlFreq2_Val2_4 }
```

**Comment by Directiveentire** [comment]
@ojwb https://stackdev.space/questions/the-undefined-identifier-enum-is-currently-not-recognized

**Comment by ojwb** [comment]
@Directiveentire I don't understand the point you are trying to make.  The issue reported here was fixed several months ago and is in 4.3.0, but was purely an issue in the Perl code SWIG was generating for this case.

If you are still seeing problems in this area with 4.3.0, please open a new ticket with text explaining the problem, not just a link to some random website (unless there's already an open ticket, e.g. #3012).
